plutus-ir/src/Language/PlutusIR.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+module Language.PlutusIR (
+    TyName (..),
+    Name (..),
+    Kind (..),
+    Type (..),
+    VarDecl (..),
+    TyVarDecl (..),
+    Datatype (..),
+    Recursivity (..),
+    Binding (..),
+    Term (..),
+    prettyDef,
+    embedIntoIR
+    ) where
+
+import           PlutusPrelude
+
+import           Language.PlutusCore        (Kind, Name, TyName, Type)
+import qualified Language.PlutusCore        as PLC
+import qualified Language.PlutusCore.Pretty as PLC
+
+import           GHC.Generics               (Generic)
+
+-- Variable declarations
+
+data VarDecl tyname name a = VarDecl a (name a) (Type tyname a)
+    deriving (Functor, Show, Eq, Generic)
+data TyVarDecl tyname a = TyVarDecl a (tyname a) (Kind a)
+    deriving (Functor, Show, Eq, Generic)
+
+-- Datatypes
+
+data Datatype tyname name a = Datatype a (TyVarDecl tyname a) [TyVarDecl tyname a] (name a) [VarDecl tyname name a]
+    deriving (Functor, Show, Eq, Generic)
+
+-- Bindings
+
+data Recursivity = NonRec | Rec
+    deriving (Show, Eq, Generic)
+data Binding tyname name a = TermBind a (VarDecl tyname name a) (Term tyname name a)
+                           | TypeBind a (TyVarDecl tyname a) (Type tyname a)
+                           | DatatypeBind a (Datatype tyname name a)
+    deriving (Functor, Show, Eq, Generic)
+
+-- Terms
+
+{- Note [Declarations in Plutus Core]
+In Plutus Core declarations are usually *flattened*, i.e. `(lam x ty t)`, whereas
+in Plutus IR they are *reified*, i.e. `(lam (vardecl x ty) t)`.
```

Comment:
Does the AST below allow `(lam (vardecl x ty) t)`? Can't see it.

Reply:
Hm yes, I should use a real example.